### DISCUSSION TITLE
Error handling in Graph

### DISCUSSION THREAD

**Discussion by avdgrinten** [networkit/networkit#769] [General]
Right now, the error handling in `Graph` is a bit messy. I'd like to have some sort of guidelines to clarify the expectations of users about what kind of error handling is done.

I propose the following:
* `get`-style functions (and related functions, such as iterators) handle all kinds of graphs (i.e., (un-)weighted/(un-)directed/(not ) indexed) correctly and return fallback values (such as edge weights of `1` and edge IDs of `none`). This even holds for unsafe functions.
* Functions marked as unsafe (only accessible from C++) ~~do not validate their inputs, i.e., passing in a node `>= G.upperNodeIdBound()` is undefined behavior~~  only validate their inputs using assertions.
* Non-unsafe functions in C++ validate their inputs and throw exceptions if inputs are out of bounds.

Any comments on this proposal?

EDIT: Proposal updated after discussion.

**Comment by michitux** [comment]
I would suggest to add asserts also for functions marked as unsafe where possible. Asserts are there to make debugging easier by validating invariants that are assumed to be true but might be violated in the case of a bug and I think that also makes sense for unsafe functions. 

What speaks against adding exceptions for non-unsafe function in C++? I think having asserts in C++ and exceptions in Python kind of duplicates these checks and adds an additional maintenance burden. I would suggest to implement all checks as C++ exceptions. In cases where this may slow down algorithms - I think this concerns in particular `addEdge()` (scanning of incident edges to avoid adding duplicate edges) and `degree(|In|Out)` (additional memory access to check `exists[u]`) - we can add an additional unsafe function that is identical to the non-unsafe function but uses asserts for those checks.

**Reply by avdgrinten** [reply]
If we add assertions to unsafe functions, I'd only enable then with `NETWORKIT_SANITY_CHECKS`. (Note that there is a CI build that builds with this flag enabled.) In contrast to `isDirected()` etc. branches, branches that check for the validity of input are much harder for the compiler to optimize (e.g., they cannot easily be hoisted out of loops), so I would rather not burden the already unsafe functions with these checks. Note that we can easily add a `checkConsistency()` call to most users of unsafe functions (when `NETWORKIT_SANITY_CHECKS` is on).

> What speaks against adding exceptions for non-unsafe function in C++?

I would rather enable assertions in release builds than throw exceptions here, at least for invalid input. Passing an incorrect node ID is really a contract violation, not a runtime error that users can act on: how are you going to handle such an exception, other than by aborting the program? (The story is different for code paths like duplicate edges - these should be handled gracefully.) Another reason not to throw exceptions in hot loops is that `throw` is an optimization barrier for the compiler, since it must ensure that side effects (e.g. to memory) are visible before the exception handler is entered (admittedly, it's not clear if this is an issue without doing extensive benchmarks first).

For Python, the expectation is different: no matter if you pass invalid input or not, you should not be able to crash the interpreter.

I think the burden of duplicating the checks is quite manageable (Graph does not have *that* many functions and many of its functions do not take input that requires validation.)

**Reply by michitux** [reply]
I've just found an argument against exceptions in methods that are called a lot in NetworKit: chances are they are called in a parallel section. And OpenMP forbids code that leaves the OpenMP block via exceptions. Even though basically any method could be called in a parallel section, I think this particularly likely for graph methods. The same is true for other data structures like `Partition` or `Cover`. Therefore, I think there is no better solution than having asserts in C++ and exceptions in Python.

Concerning unsafe functions, I have always assumed that asserts being enabled basically means that performance is not relevant, similar to `NETWORKIT_SANITY_CHECKS`, and that asserts are therefore not enabled in release builds. If you want to make a difference here, I don't mind, but this should be clearly documented and existing code should be checked to ensure that asserts are cheap.

**Comment by hmeyerhenke** [comment]
When you say "handle all kinds of graphs", you only refer to the Graph class itself, right? Because most algorithms are not made for all kinds of graphs.

**Reply by avdgrinten** [reply]
Yes.

**Comment by michitux** [comment]
This is also not only an issue in the graph class, this also concerns other classes. For example, I'm aware that quality measures like `Modularity` and distance measures like `AdjustedRandMeasure` crash if the `Partition`(s) and `Graph` do not have the same number of nodes. They also crash if the upper bound is not correctly set on the Partition(s) (i.e., if there are nodes that have a higher partition id set or even `none` as partition id of a node). Some (most?) of these cases are handled via asserts. Just enabling asserts in release builds doesn't help here, as this still allows to crash the Python interpreter. Either those asserts need to be turned into exceptions, or the checks need to be duplicated in Python. Therefore, I think we need a general policy here what is expected from algorithms and data structures, not just for the graph class.

**Reply by avdgrinten** [reply]
For these checks, I have no issues with exceptions, I just don't want to handle sanity checking in hot loops via exceptions.

**Reply by avdgrinten** [reply]
And I agree: it is not feasible to duplicate these checks in C++ and Python (so assertion in C++ are indeed not attractive here, not even if they are enabled in debugging builds).

**Reply by michitux** [reply]
Note that most of these existing asserts in the above-mentioned algorithms are in hot loops. Those algorithms consider it a violation of contract if the given partition and graph do not match or the partition contains `none` values for existing nodes. I don't see the reason why those algorithms should use exceptions while the graph uses asserts.

**Reply by avdgrinten** [reply]
Can you point me to an example for such a check?

As for why I want to clarify the expectations for the Graph class first: the Graph class is used by *every* algorithm, so picking a strategy with high performance overhead affects the entire toolkit. Note that in this thread, I am mostly concerned with low-level operations that are called frequently (i.e., those that you usually do not want to call from Python unless absolutely necessary); if high-level operations (such as the ones that operate on entire graphs etc., or Algorithm::run()) have a different error reporting mechanism, I'm totally fine with that.

**Reply by michitux** [reply]
Have a look at [NMIDistance.cpp#L28](https://github.com/networkit/networkit/blob/35fbf23292223d6173aaa7f008f1e16804befc49/networkit/cpp/community/NMIDistance.cpp#L28) or [Modularity.cpp#L45](https://github.com/networkit/networkit/blob/35fbf23292223d6173aaa7f008f1e16804befc49/networkit/cpp/community/Modularity.cpp#L45). Partition ids can be any value, there is no way in the partition class to prevent that arbitrary values are set as writable references to them are returned by the `[]`-operator. Note that these algorithms are definitely missing some more checks or contain wrong checks, e.g., `Modularity::getQuality()` checks `assert (G.numberOfNodes() <= zeta.numberOfElements());` but should check `assert (G.uperNodeIdBound() == zeta.numberOfElements());`. In fact, I wouldn't be surprised if (almost) all algorithms that take a partition or a (set of) node(s) as input did not validate their inputs (properly), not even with asserts. Even my new local community detection algorithms in #448 only check by chance if the given nodes are actually existing nodes, e.g., by calling a graph method that contains a check.

**Comment by hmeyerhenke** [comment]
Sounds good to me.

**Comment by avdgrinten** [comment]
Alright, after the discussion I agree that exceptions are better than assertions + special handling at the Python side, so let's change this proposal to throw exceptions in non-unsafe methods and implement it.
